src/Update/RecipePatcher.php

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+    }
+
+    public function generatePatch(array $originalFiles, array $newFiles): RecipePatch
+    {
+        // null implies "file does not exist"
+        $originalFiles = array_filter($originalFiles, function ($file) {
+            return null !== $file;
+        });
+        $newFiles = array_filter($newFiles, function ($file) {
+            return null !== $file;
+        });
+
+        // find removed files and add them so they will be deleted
+        foreach ($originalFiles as $file => $contents) {
+            if (!isset($newFiles[$file])) {
+                $newFiles[$file] = null;
+            }
+        }
+
+        // If a file is being modified, but does not exist in the current project,
+        // it cannot be patched. Ignore these files and leave them deleted.
+        $modifiedFiles = array_intersect_key(array_keys($originalFiles), array_keys($newFiles));
+        foreach ($modifiedFiles as $modifiedFile) {
+            if (!file_exists($this->rootDir.'/'.$modifiedFile)) {
+                unset($originalFiles[$modifiedFile]);
+                unset($newFiles[$modifiedFile]);
+            }
+        }
+
+        $tmpPath = sys_get_temp_dir().'/_flex_recipe_update'.uniqid(mt_rand(), true);
+        $this->filesystem->mkdir($tmpPath);
+
+        try {
+            $this->execute('git init', $tmpPath);
+            $this->execute('git config user.name "Flex Updater"', $tmpPath);
+            $this->execute('git config user.email ""', $tmpPath);
+
+            $blobs = [];
+            if (\count($originalFiles) > 0) {
+                $this->writeFiles($originalFiles, $tmpPath);
+                $this->execute('git add -A', $tmpPath);
+                $this->execute('git commit -m "original files"', $tmpPath);
+
+                $blobs = $this->generateBlobs($originalFiles, $tmpPath);
+            }
+
+            $this->writeFiles($newFiles, $tmpPath);
+            $this->execute('git add -A', $tmpPath);
+
+            return new RecipePatch(
+                $this->execute('git diff --cached', $tmpPath),
+                $blobs
+            );
+        } finally {
+            try {
+                $this->filesystem->remove($tmpPath);
+            } catch (IOException $e) {
+                // this can sometimes fail due to git file permissions
+                // if that happens, just leave it: we're in the temp directory anyways
+            }
+        }
+    }
+
+    private function writeFiles(array $files, string $directory): void
+    {
+        foreach ($files as $filename => $contents) {
+            $path = $directory.'/'.$filename;
+            if (null === $contents) {
+                if (file_exists($path)) {
+                    unlink($path);
+                }
+
+                continue;
+            }
+
+            if (!file_exists(\dirname($path))) {
+                $this->filesystem->mkdir(\dirname($path));
+            }
+            file_put_contents($path, $contents);
```

Comment:
if content is null (case when file exists in `original` but not in `new`) should it delete the file instead of putting an empty content in it?

Reply:
It does this a few lines up - there's a check for null `$contents` and an `unlink` (if the file exists - because if this is the starting file, then there is nothing yet to delete). So unless you're seeing something that looks funny to you, it already deletes on null :).